pkg/scheduler/framework/framework.go

```diff
@@ -143,8 +174,738 @@ func (f *framework) EventRecorder() record.EventRecorder {
+	// changes eventually.
+	clusters, err := f.collectClusters(ctx)
+	if err != nil {
+		klog.ErrorS(err, "failed to collect clusters", "schedulingPolicySnapshot", schedulingPolicySnapshotRef)
+		return ctrl.Result{}, err
+	}
+
+	// Collect all bindings.
+	//
+	// Note that for consistency reasons, bindings are listed directly from the API server; this helps
+	// avoid a classic read-after-write consistency issue, which, though should only happen when there
+	// are connectivity issues and/or API server is overloaded, can lead to over-scheduling in adverse
+	// scenarios. It is true that even when bindings are over-scheduled, the scheduler can still correct
+	// the situation in the next cycle; however, considering that placing resources to clusters, unlike
+	// pods to nodes, is more expensive, it is better to avoid over-scheduling in the first place.
+	//
+	// This, of course, has additional performance overhead (and may further exacerbate API server
+	// overloading). In the long run we might still want to resort to a cached situtation.
+	//
+	// TO-DO (chenyu1): explore the possbilities of using a mutation cache for better performance.
+	bindings, err := f.collectBindings(ctx, crpName)
+	if err != nil {
+		klog.ErrorS(err, "failed to collect bindings", "schedulingPolicySnapshot", schedulingPolicySnapshotRef)
+		return ctrl.Result{}, err
+	}
+
+	// Parse the bindings, find out
+	//
+	// * active bindings, i.e., bindings that are associated with a normally operating cluster and
+	//   have been cleared for processing by the dispatcher; and
+	// * creating bindings, i.e., bindings that have been associated with a normally operating cluster,
+	//   but have not yet been cleared for processing by the dispatcher; and
+	// * obsolete bindings, i.e., bindings that are scheduled in accordance with an out-of-date
+	//   (i.e., no longer active) scheduling policy snapshot; it may or may have been cleared for
+	//   processing by the dispatcher; and
+	// * dangling bindings, i.e., bindings that are associated with a cluster that is no longer
+	//   in a normally operating state (the cluster has left the fleet, or is in the state of leaving),
+	//   yet has not been marked as deleting by the scheduler; and
+	// * deleted bindings, i.e, bindings that are marked for deletion in the API server, but have not
+	//   yet been marked as deleting by the scheduler.
+	//
+	// Note that bindings marked as deleting are ignored by the scheduler (unless they are already marked
+	// for deletion and still has the scheduler cleanup finalizer), as they
+	// are irrelevant to the scheduling cycle.
+	active, creating, obsolete, dangling, deleted := classifyBindings(policy, bindings, clusters)
+
+	// If a binding has been marked for deletion yet still has the scheduler cleanup finalizer,
+	// remove the finalizer from it to clear it for GC.
+	if err := f.removeSchedulerCleanupFinalizerFrom(ctx, deleted); err != nil {
```

Comment:
since the rollout controller issue the real delete request to api-server. I wonder why does the scheduler need to handle this?

Reply:
Hi Ryan! The scheduler finalizer is added to make sure that the scheduler will be able to catch any binding being deleted (even if it is manually deleted) (and consequently correct the situation).